From 34d703be14ab2a9741dd10ac4e54d906ec3328a1 Mon Sep 17 00:00:00 2001
From: Kazunori Kobayashi <kkobayas@igel.co.jp>
Date: Mon, 15 Apr 2013 12:12:07 +0900
Subject: [PATCH 594/611] ARM: shmobile: r8a7778: add ICB MERAM work-around

r8a7778 chip has errata that causes an incorrect read access
from VPU5HD2 to ICB. This change is workaround against this issue.

The workaround has to access non-MERAM register address, so it
can't be controlled with MERAM driver. This change adds a function
that sets the workaround to setup-r8a7778.c. It will be invoked
from board specific code.
---
 arch/arm/mach-shmobile/include/mach/common.h |  2 ++
 arch/arm/mach-shmobile/setup-r8a7778.c       | 16 ++++++++++++++++
 2 files changed, 18 insertions(+)

diff --git a/arch/arm/mach-shmobile/include/mach/common.h b/arch/arm/mach-shmobile/include/mach/common.h
index 866e9f4..3088f6b 100644
--- a/arch/arm/mach-shmobile/include/mach/common.h
+++ b/arch/arm/mach-shmobile/include/mach/common.h
@@ -48,6 +48,8 @@ extern void r8a7740_clock_init(u8 md_ck);
 extern void r8a7740_reserve_memory(void);
 extern void r8a7740_pinmux_init(void);
 
+extern void r8a7778_meram_workaround(void);
+
 extern void r8a7779_init_irq(void);
 extern void r8a7779_init_irq_extpin(int irlm);
 extern void r8a7779_init_irq_dt(void);
diff --git a/arch/arm/mach-shmobile/setup-r8a7778.c b/arch/arm/mach-shmobile/setup-r8a7778.c
index 75ad38a..fa6d6cb 100644
--- a/arch/arm/mach-shmobile/setup-r8a7778.c
+++ b/arch/arm/mach-shmobile/setup-r8a7778.c
@@ -671,3 +671,19 @@ DT_MACHINE_START(R8A7778_DT, "Generic R8A7778 (Flattened Device Tree)")
 MACHINE_END
 
 #endif /* CONFIG_USE_OF */
+
+/*
+ * r8a7778 chip has lasting errata on MERAM buffer.
+ * this is work-around for it.
+ */
+#define MEPRICNT	0xfd0500ac
+void r8a7778_meram_workaround(void)
+{
+	void __iomem *reg;
+
+	reg = ioremap_nocache(MEPRICNT, 4);
+	if (reg) {
+		iowrite32(0x0f880f0f, reg);
+		iounmap(reg);
+	}
+}
-- 
1.8.2

